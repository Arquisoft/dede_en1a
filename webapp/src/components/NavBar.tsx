import ShoppingCartIcon from "@mui/icons-material/ShoppingCart";
import {Badge} from '@mui/material';
import React, {useEffect, useState} from 'react'
import {useContext} from 'react';
import {Container, Nav, Navbar} from 'react-bootstrap';
import {Link} from 'react-router-dom';
import {CartContext} from '../context/CartContext';
import {getTotalItems} from '../helpers/calculate';
import {IconButton} from "@mui/material";
import LoginButtonSolid from "./login/LoginButtonComponent";
import {LogoutButtonSolid} from "./logout/LogoutButtonComponent";
import {useSession} from "@inrupt/solid-ui-react"

type Props = {
    handleOpen: (state: boolean) => void;
}

const NavBar = ({handleOpen}: Props) => {
    const {cartItems} = useContext(CartContext);
    const [isLoggedIn, setIsLoggedIn] = useState(false)
    const {session} = useSession()



    useEffect(() => {
        if(isLoggedIn && session.info.webId)
            localStorage.setItem("webId", session.info.webId)
        else
            localStorage.removeItem("webId")
    }, [isLoggedIn, session.info.webId])

    session.onLogin(() => {
        setIsLoggedIn(true)
    })

    session.onLogout(() => {
        setIsLoggedIn(false)
    })

    //We have logged in, help
    return (
        <Navbar bg="dark" variant='dark' expand="lg">
            <Container>
                <Link className='navbar-brand' to='/'>
                    <img src='/images/logo.png' height='45px' width='auto' alt="logo"/>
                </Link>
                <Navbar.Toggle aria-controls="basic-navbar-nav"/>
                <Navbar.Collapse id="basic-navbar-nav">
                    <Nav className="mr-auto">
                        <Link className='nav-link' to="/">Home</Link>
                    </Nav>
                    <Nav className="mr-auto">
                        <Link className='nav-link' to="/orders/list" hidden={!isLoggedIn}>My orders</Link>
                    </Nav>
                </Navbar.Collapse>
                <IconButton
                    size="large"
                    color="primary"
                    onClick={() => handleOpen(true)}
                    sx={{mr: 2}}
                >
                    <Badge badgeContent={getTotalItems(cartItems)} color="error">
                        <ShoppingCartIcon />
                    </Badge>
                </IconButton>
                {!isLoggedIn ? <LoginButtonSolid/> : <LogoutButtonSolid/>}
            </Container>
        </Navbar>
    )
}

export default NavBar;