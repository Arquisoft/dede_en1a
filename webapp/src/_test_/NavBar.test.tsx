import React from 'react';
import ReactDOM from 'react-dom';
import {BrowserRouter as Router} from "react-router-dom";
import { render, screen, cleanup } from '@testing-library/react';
import NavBar from '../components/NavBar';

const handleOpen = jest.fn(() => true)

test("NavBar contains logo", () => {
    const fakeNavBar = render(<Router><NavBar isLoggedIn={false} handleOpen={handleOpen}/></Router>)
    //expect(fakeNavBar).toContain(logo); go back and fix this
})

test("NavBar matches snapshot", () => {
    const fakeNavBar = render(<Router><NavBar isLoggedIn={false} handleOpen={handleOpen}/></Router>)
    expect(fakeNavBar).toMatchSnapshot();
})

// test distinct components in separate test scripts?
// test that buttons work correctly?
