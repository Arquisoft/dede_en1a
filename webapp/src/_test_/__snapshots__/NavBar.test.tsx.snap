// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`NavBar matches snapshot 1`] = `
Object {
  "asFragment": [Function],
  "baseElement": <body>
    <div>
      <nav
        class="navbar navbar-expand-lg navbar-dark bg-dark"
      >
        <div
          class="container"
        >
          <a
            class="navbar-brand"
            href="/"
          >
            DEDE
          </a>
          <button
            aria-controls="basic-navbar-nav"
            aria-label="Toggle navigation"
            class="navbar-toggler collapsed"
            type="button"
          >
            <span
              class="navbar-toggler-icon"
            />
          </button>
          <div
            class="navbar-collapse collapse"
            id="basic-navbar-nav"
          >
            <div
              class="mr-auto navbar-nav"
            >
              <a
                class="nav-link"
                href="/"
              >
                Home
              </a>
            </div>
          </div>
          <button
            class="MuiButtonBase-root MuiIconButton-root MuiIconButton-colorPrimary MuiIconButton-sizeLarge css-4zrcd2-MuiButtonBase-root-MuiIconButton-root"
            id="shoppingCart"
            tabindex="0"
            type="button"
          >
            <span
              class="MuiBadge-root css-1c32n2y-MuiBadge-root"
            >
              <svg
                aria-hidden="true"
                class="MuiSvgIcon-root MuiSvgIcon-fontSizeMedium css-i4bv87-MuiSvgIcon-root"
                data-testid="ShoppingCartIcon"
                focusable="false"
                viewBox="0 0 24 24"
              >
                <path
                  d="M7 18c-1.1 0-1.99.9-1.99 2S5.9 22 7 22s2-.9 2-2-.9-2-2-2zM1 2v2h2l3.6 7.59-1.35 2.45c-.16.28-.25.61-.25.96 0 1.1.9 2 2 2h12v-2H7.42c-.14 0-.25-.11-.25-.25l.03-.12.9-1.63h7.45c.75 0 1.41-.41 1.75-1.03l3.58-6.49c.08-.14.12-.31.12-.48 0-.55-.45-1-1-1H5.21l-.94-2H1zm16 16c-1.1 0-1.99.9-1.99 2s.89 2 1.99 2 2-.9 2-2-.9-2-2-2z"
                />
              </svg>
              <span
                class="MuiBadge-anchorOriginTopRightRectangular MuiBadge-overlapRectangular MuiBadge-colorError MuiBadge-badge MuiBadge-standard MuiBadge-anchorOriginTopRight MuiBadge-invisible css-zkqb44-MuiBadge-badge"
              />
            </span>
            <span
              class="MuiTouchRipple-root css-8je8zh-MuiTouchRipple-root"
            />
          </button>
          <button
            class="MuiButton-root MuiButton-text MuiButton-textPrimary MuiButton-sizeMedium MuiButton-textSizeMedium MuiButtonBase-root  css-1e6y48t-MuiButtonBase-root-MuiButton-root"
            tabindex="0"
            type="button"
          >
            Login
            <span
              class="MuiTouchRipple-root css-8je8zh-MuiTouchRipple-root"
            />
          </button>
        </div>
      </nav>
    </div>
  </body>,
  "container": <div>
    <nav
      class="navbar navbar-expand-lg navbar-dark bg-dark"
    >
      <div
        class="container"
      >
        <a
          class="navbar-brand"
          href="/"
        >
          DEDE
        </a>
        <button
          aria-controls="basic-navbar-nav"
          aria-label="Toggle navigation"
          class="navbar-toggler collapsed"
          type="button"
        >
          <span
            class="navbar-toggler-icon"
          />
        </button>
        <div
          class="navbar-collapse collapse"
          id="basic-navbar-nav"
        >
          <div
            class="mr-auto navbar-nav"
          >
            <a
              class="nav-link"
              href="/"
            >
              Home
            </a>
          </div>
        </div>
        <button
          class="MuiButtonBase-root MuiIconButton-root MuiIconButton-colorPrimary MuiIconButton-sizeLarge css-4zrcd2-MuiButtonBase-root-MuiIconButton-root"
          id="shoppingCart"
          tabindex="0"
          type="button"
        >
          <span
            class="MuiBadge-root css-1c32n2y-MuiBadge-root"
          >
            <svg
              aria-hidden="true"
              class="MuiSvgIcon-root MuiSvgIcon-fontSizeMedium css-i4bv87-MuiSvgIcon-root"
              data-testid="ShoppingCartIcon"
              focusable="false"
              viewBox="0 0 24 24"
            >
              <path
                d="M7 18c-1.1 0-1.99.9-1.99 2S5.9 22 7 22s2-.9 2-2-.9-2-2-2zM1 2v2h2l3.6 7.59-1.35 2.45c-.16.28-.25.61-.25.96 0 1.1.9 2 2 2h12v-2H7.42c-.14 0-.25-.11-.25-.25l.03-.12.9-1.63h7.45c.75 0 1.41-.41 1.75-1.03l3.58-6.49c.08-.14.12-.31.12-.48 0-.55-.45-1-1-1H5.21l-.94-2H1zm16 16c-1.1 0-1.99.9-1.99 2s.89 2 1.99 2 2-.9 2-2-.9-2-2-2z"
              />
            </svg>
            <span
              class="MuiBadge-anchorOriginTopRightRectangular MuiBadge-overlapRectangular MuiBadge-colorError MuiBadge-badge MuiBadge-standard MuiBadge-anchorOriginTopRight MuiBadge-invisible css-zkqb44-MuiBadge-badge"
            />
          </span>
          <span
            class="MuiTouchRipple-root css-8je8zh-MuiTouchRipple-root"
          />
        </button>
        <button
          class="MuiButton-root MuiButton-text MuiButton-textPrimary MuiButton-sizeMedium MuiButton-textSizeMedium MuiButtonBase-root  css-1e6y48t-MuiButtonBase-root-MuiButton-root"
          tabindex="0"
          type="button"
        >
          Login
          <span
            class="MuiTouchRipple-root css-8je8zh-MuiTouchRipple-root"
          />
        </button>
      </div>
    </nav>
  </div>,
  "debug": [Function],
  "findAllByAltText": [Function],
  "findAllByDisplayValue": [Function],
  "findAllByLabelText": [Function],
  "findAllByPlaceholderText": [Function],
  "findAllByRole": [Function],
  "findAllByTestId": [Function],
  "findAllByText": [Function],
  "findAllByTitle": [Function],
  "findByAltText": [Function],
  "findByDisplayValue": [Function],
  "findByLabelText": [Function],
  "findByPlaceholderText": [Function],
  "findByRole": [Function],
  "findByTestId": [Function],
  "findByText": [Function],
  "findByTitle": [Function],
  "getAllByAltText": [Function],
  "getAllByDisplayValue": [Function],
  "getAllByLabelText": [Function],
  "getAllByPlaceholderText": [Function],
  "getAllByRole": [Function],
  "getAllByTestId": [Function],
  "getAllByText": [Function],
  "getAllByTitle": [Function],
  "getByAltText": [Function],
  "getByDisplayValue": [Function],
  "getByLabelText": [Function],
  "getByPlaceholderText": [Function],
  "getByRole": [Function],
  "getByTestId": [Function],
  "getByText": [Function],
  "getByTitle": [Function],
  "queryAllByAltText": [Function],
  "queryAllByDisplayValue": [Function],
  "queryAllByLabelText": [Function],
  "queryAllByPlaceholderText": [Function],
  "queryAllByRole": [Function],
  "queryAllByTestId": [Function],
  "queryAllByText": [Function],
  "queryAllByTitle": [Function],
  "queryByAltText": [Function],
  "queryByDisplayValue": [Function],
  "queryByLabelText": [Function],
  "queryByPlaceholderText": [Function],
  "queryByRole": [Function],
  "queryByTestId": [Function],
  "queryByText": [Function],
  "queryByTitle": [Function],
  "rerender": [Function],
  "unmount": [Function],
}
`;
